python/tvm/contrib/pipeline_executor.py

```diff
@@ -0,0 +1,395 @@
+    dependency relation between Interfaces this class can build the module
+    connection relation.
+
+    The class Hierarchical as following.
+         PipelineConfig ---> ModuleWrapper ---> Interface(input/output)
+    """
+
+    class ModuleWrapper:
+        """The class use use to represent Module and storage module index and
+        Interface information.
+        """
+
+        class Interface:
+            """The class that use to storage module connection information.
+               There are 2 types Interface Input:1 Output:2
+            Parameters
+            ----------
+
+            owner : ModuleWrapper
+                The class that own this interface, in such class there are
+                Module information like index, module name
+
+            itype : integer
+                Interface type, 1 is input interface, 2 is output interface
+
+            name : str/integer
+                Interface name, for input that is string for example "data0"
+                for output that is integer for example 0.
+            """
+
+            def __init__(self, owner, itype, name):
+                self.owner_ = owner
+                self.itype_ = itype
+                self.name_ = str(name)
+                self.dependent_ = []
+
+            def get_name(self):
+                mname = ""
+                if self.owner_:
+                    mname = self.owner_.name_
+
+                return mname, self.name_
+
+            def get_owner_indx(self):
+                return self.owner_.indx_
+
+            def get_dependent_str(self):
+                name = ""
+                for dependent in self.dependent_:
+                    mname, dname = dependent.get_name()
+                    name = name + (mname + ":output(" + dname if self.itype_ == 2 else "")
+                    name = name + (")" if self.itype_ == 2 else mname + ":" + dname)
+                return name
+
+            def add_dependent(self, dependent):
+                """
+                # check if the dependency setting correct.
+                # correct connection are following
+                # 1. global input to module input
+                # 2. module output to next module input
+                # 3. module output to global output
+                """
+                owner_indx = self.get_owner_indx()
+                dep_owner_indx = dependent.get_owner_indx()
+                assert owner_indx != dep_owner_indx, f"can not set self as dependent."
+                assert not (
+                    owner_indx > dep_owner_indx
+                    and not (dependent.itype_ == 2 and dep_owner_indx == 0)
+                ), f"dependent only can be next module interface or global output."
+                assert not (
+                    owner_indx == 0 and dependent.itype_ != 1
+                ), f"global input only can set dependent with module input."
+
+                self.dependent_.append(dependent)
```

Comment:
In general you need to raise exceptions instead of `assert` in this function. `assert` is usually used for internal error that developers should be aware of. In this case, these errors are triggered by incorrect user behaviors.

Meanwhile, there are two points for the user message:
1. Use capital in the first word.
2. No need to add a period at the end.

```suggestion
            def add_binding(self, binding):
                """Check if the binding is legal by looking at the following:
                1. Cannot bind to the module itself.
                2. Can only bind to later stage modules or global outputs.
                3. module output to global output.
                """
                owner_indx = self.get_owner_indx()
                dep_owner_indx = binding.get_owner_indx()
                if owner_indx == dep_owner_indx:
                    raise RuntimeError(f"Illegal binding: {self.name} cannot bind to the same module")
                if owner_indx < dep_owner_indx:
                    raise Runtimeerror(f"Illegal binding: {self.name} cannot bind to {binding.name} before it")              
                if dependent.io_type == "output" and dep_owner_indx == 0: # Cannot understand this rule.
                    raise RuntimeError(f"...")
                if owner_indx == 0 and binding.io_type != "input": # Cannot understand this rule either.
                    raise RuntimeError(f"...")

                self.bindings.append(binding)
```

I'm not sure if I changed the conditions correctly as I cannot understand some of them. Please check if the conditions are correct or not.

Reply:
about owner_indx == 0 and binding.io_type != "input", here define a rule that the connection/binding should from top to down, pipe_config["input"] --> module["input"] --> pipe_config["ouput"], that means here can not start from pipe_config["output"]["0"] to do any connect/binding.

about not (dependent.itype_ == 2 and dep_owner_indx == 0),  this means the target of connect/binding only can be either next module or gloabl output

Reply:
changed into raise RuntimeError